nursery/coin_select/src/coin_selector.rs

```diff
@@ -176,440 +174,424 @@ impl<'a> CoinSelector<'a> {
+        let weight = TXIN_BASE_WEIGHT + satisfaction_weight;
+        Candidate {
+            value,
+            weight,
+            input_count: 1,
+            is_segwit,
         }
     }
-}
 
-#[derive(Clone, Debug)]
-pub struct Selection {
-    pub selected: BTreeSet<usize>,
-    pub excess: u64,
-    pub excess_strategies: HashMap<ExcessStrategyKind, ExcessStrategy>,
-}
+    /// Effective value of this input candidate: `actual_value - input_weight * feerate (sats/wu)`.
+    pub fn effective_value(&self, feerate: FeeRate) -> Ordf32 {
+        Ordf32(self.value as f32 - (self.weight as f32 * feerate.spwu()))
+    }
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, core::hash::Hash)]
-pub enum ExcessStrategyKind {
-    ToFee,
-    ToRecipient,
-    ToDrain,
+    /// Value per weight unit
+    pub fn value_pwu(&self) -> Ordf32 {
+        Ordf32(self.value as f32 / self.weight as f32)
+    }
 }
 
-#[derive(Clone, Copy, Debug)]
-pub struct ExcessStrategy {
-    pub recipient_value: Option<u64>,
-    pub drain_value: Option<u64>,
-    pub fee: u64,
+/// A drain (A.K.A. change) output.
+/// Technically it could represent multiple outputs.
+///
+/// These are usually created by a [`change_policy`].
+///
+/// [`change_policy`]: crate::change_policy
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
+pub struct Drain {
+    /// The weight of adding this drain
     pub weight: u32,
-    pub waste: i64,
+    /// The value that should be assigned to the drain
+    pub value: u64,
+    /// The weight of spending this drain
+    pub spend_weight: u32,
 }
 
-impl Selection {
-    pub fn apply_selection<'a, T>(
-        &'a self,
-        candidates: &'a [T],
-    ) -> impl Iterator<Item = &'a T> + 'a {
-        self.selected.iter().map(move |i| &candidates[*i])
+impl Drain {
+    /// A drian representing no drain at all.
+    pub fn none() -> Self {
+        Self::default()
     }
 
-    /// Returns the [`ExcessStrategy`] that results in the least waste.
-    pub fn best_strategy(&self) -> (&ExcessStrategyKind, &ExcessStrategy) {
-        self.excess_strategies
-            .iter()
-            .min_by_key(|&(_, a)| a.waste)
-            .expect("selection has no excess strategy")
+    /// is the "none" drain
+    pub fn is_none(&self) -> bool {
+        self == &Drain::none()
     }
-}
 
-impl core::fmt::Display for ExcessStrategyKind {
-    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        match self {
-            ExcessStrategyKind::ToFee => core::write!(f, "to_fee"),
-            ExcessStrategyKind::ToRecipient => core::write!(f, "to_recipient"),
-            ExcessStrategyKind::ToDrain => core::write!(f, "to_drain"),
+    /// Is not the "none" drain
+    pub fn is_some(&self) -> bool {
+        !self.is_none()
+    }
+
+    pub fn new_tr_keyspend() -> Self {
+        Self {
+            weight: TXOUT_BASE_WEIGHT + TR_SPK_WEIGHT,
```

Comment:
Yep well spotted. I added a 1 byte spk length to `TXOUT_BASE_WEIGHT`. All this weight stuff was added on a whim while I was trying to use this and is not very well tested. I'm not sure if it's worth having since you should probably get the numbers from rust bitcoin and rust miniscript which do a better job of telling you this stuff now.